### PR TITLE
Move sync logic to contoller

### PR DESCRIPTION
**Pull request by alexmt**
Implementation is not completed yet: tests, more comments and rebase is requred. @jessesuen , @merenbach can you please give your feedback for this initial version.

### CODE REVIEW

File: reposerver/repository/repository.proto [hunk 1]
```diff
@@ -23,17 +23,7 @@ message ManifestResponse {
...
-message EnvParamsRequest {
-    github.com.argoproj.argo_cd.pkg.apis.application.v1alpha1.Repository repo = 1;
-    string revision = 2;
-    string path = 3;
-    string environment = 4;
-}
-
-message EnvParamsResponse {
-    repeated github.com.argoproj.argo_cd.pkg.apis.application.v1alpha1.ComponentParameter params = 1;
+    repeated github.com.argoproj.argo_cd.pkg.apis.application.v1alpha1.ComponentParameter params = 5;
```
Comment (jessesuen) [inline reposerver/repository/repository.proto]: Nice change!

File: server/application/application.go [hunk 2]
```diff
@@ -498,134 +383,60 @@ func (s *Server) getRepo(ctx context.Context, repoURL string) *appv1.Repository
...
-			}
-		}
-		syncRes.Resources = append(syncRes.Resources, &resDetails)
+func (s *Server) setAppOperation(ctx context.Context, appName string, operationCreator func(app *appv1.Application) (*appv1.Operation, error)) (*appv1.Application, error) {
+	app, err := s.Get(ctx, &ApplicationQuery{Name: appName})
+	if err != nil {
+		return nil, err
+	}
+	if app.Operation != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "other operation is already in progress")
```
Comment (jessesuen) [inline server/application/application.go]: lets use the message: "another operation is already in progress".

One thing to think about is possibly making this idempotent. If the operation is identical, then we can return success. Not sure if this is truly correct behavior though.

File: util/cache/redis.go [hunk 3]
```diff
@@ -45,5 +45,5 @@ func (r *redisCache) Get(key string, obj interface{}) error {
 	if err == rediscache.ErrCacheMiss {
 		return ErrCacheMiss
 	}
-	return nil
+	return err
```
Comment (jessesuen) [inline util/cache/redis.go]: Thank you!

**Review by merenbach — APPROVED** [review]

File: pkg/apis/application/v1alpha1/types.go [hunk 4]
```diff
@@ -11,6 +11,48 @@ import (
...
+	Revision    string               `json:"revision" protobuf:"bytes,1,opt,name=revision"`
+	Prune       bool                 `json:"prune" protobuf:"bytes,2,opt,name=prune"`
+	Overrides   []ComponentParameter `json:"overrides,omitempty" protobuf:"bytes,3,opt,name=overrides"`
+	NoOverrides bool                 `json:"noOverrides,omitempty" protobuf:"bytes,4,opt,name=noOverrides"`
+	DryRun      bool                 `json:"dryRun" protobuf:"bytes,5,opt,name=dryRun"`
+}
+
+// Operation contains requested operation parameters.
+type Operation struct {
+	Sync *SyncOperation `json:"sync" protobuf:"bytes,2,opt,name=sync"`
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: As discussed, lets have RollbackOperation as a second field and get rid of NoOverrides. Also why did the protobuf start at 2?

File: pkg/apis/application/v1alpha1/generated.proto [hunk 5]
```diff
@@ -177,6 +181,20 @@ message HealthStatus {
...
+  optional SyncOperation sync = 2;
+}
+
+// OperationState contains information about state of currently performing operation on application.
+message OperationState {
+  optional string status = 1;
+
+  optional string errorDetails = 2;
+
+  optional SyncOperationResult sync = 3;
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/generated.proto]: I think we may need RollbackOperationResult as well, if we have separate operations.

File: controller/health.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	clusterService cluster.ClusterServiceServer
+	namespace      string
+}
+
+func NewAppHealthManager(clusterService cluster.ClusterServiceServer,
+	namespace string) AppHealthManager {
+	return &kubeAppHealthManager{clusterService: clusterService, namespace: namespace}
+}
+
+func (ctrl *kubeAppHealthManager) getServiceHealth(config *rest.Config, namespace string, name string) (*appv1.HealthStatus, error) {
```
[the remark this answers is not in the document]
Comment (alexmt) [reply controller/health.go]: I would prefer to do it in separate PR. Created ticket to track it: https://github.com/argoproj/argo-cd/issues/192

File: pkg/apis/application/v1alpha1/generated.proto [hunk 7]
```diff
@@ -215,6 +243,26 @@ message ResourceState {
...
 
+// SyncOperation contains sync operation details.
+message SyncOperation {
+  optional string revision = 1;
+
+  optional bool prune = 2;
+
+  repeated ComponentParameter overrides = 3;
+
+  optional bool noOverrides = 4;
```
[the remark this answers is not in the document]
Comment (alexmt) [reply pkg/apis/application/v1alpha1/generated.proto]: Field is not required after added separate rollback operation

File: server/application/application.go (hunk 2 above)
Comment (alexmt) [reply server/application/application.go]: Updated message.

If app resources were deleted manually then user want to sync even if the previous sync operation is identical and successful.

File: util/cache/redis.go (hunk 3 above)
Comment (alexmt) [reply util/cache/redis.go]: 👍

File: pkg/apis/application/v1alpha1/types.go (hunk 4 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Done. Fixed proto number, thank you!

File: pkg/apis/application/v1alpha1/generated.proto (hunk 5 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/generated.proto]: Done

**Comment by alexmt** [comment]
Thank you for review @merenbach , @jessesuen . Apply reviewer notes and finally got e2e tests working. PTAL

**Review by jessesuen — APPROVED** [review]

**Pull request merged**